lldb/source/Plugins/SymbolFile/DWARF/DWARFUnit.cpp

```diff
@@ -663,6 +663,14 @@ DWARFUnit::GetDIE(dw_offset_t die_offset) {
   return DWARFDIE(); // Not found
 }
 
+llvm::StringRef DWARFUnit::PeekDIEName(dw_offset_t die_offset) {
+  const DWARFDataExtractor &data = GetData();
+  DWARFDebugInfoEntry die;
+  if (!die.Extract(data, this, &die_offset))
+    return llvm::StringRef();
+  return die.GetName(this);
+}
```

Comment:
One option here is to check if all of the DIEs have been already parsed in the DWARFUnit and if so, grab the existing DIE, and if it hasn't been parsed do this. 

This approach doesn't take care of cases where the current DIE doesn't have a `DW_AT_name`, but it has a `DW_AT_specification` or `DW_AT_abstract_origin` attribute that points to another DIE that does have a name. In this case we will get no name back from this function because it doesn't recurse into the `DW_AT_specification` or `DW_AT_abstract_origin` die.

Reply:
> One option here is to check if all of the DIEs have been already parsed in the DWARFUnit and if so, grab the existing DIE, and if it hasn't been parsed do this.

Correct me if I am wrong, but even if we take an already parsed DIE, won't we be doing the exact same thing? It eventually calls `DWARFDebugInfoEntry::GetAttributeValue`, which is a more complicated way of doing the above (but it still creates the extractor, etc...)


> This approach doesn't take care of cases where the current DIE doesn't have a DW_AT_name, but it has a DW_AT_specification or DW_AT_abstract_origin

Oh great point, we need this for sure! I will update the implementation and add tests for it

Reply:
Actually, the current approach should take care of that already; we call `DWARFDebugInfoEntry::GetName`, whose implementation is:

```
const char *DWARFDebugInfoEntry::GetName(const DWARFUnit *cu) const {
  return GetAttributeValueAsString(cu, DW_AT_name, nullptr, true);
}
```

This is going to look through AT_specification and AT_abstract_origin.

I should still add a  test to verify this though!

Reply:
If the attribute isn't in the current DIE (it is in a `DW_AT_specification` or `DW_AT_abstract_origin`), the `DWARFDebugInfoEntry::GetAttributeValue(...)` function will just run this code:
```
    if (GetAttributeValue(cu, DW_AT_specification, form_value)) {
      DWARFDIE die = form_value.Reference();
      if (die) {
```
or
```
    if (GetAttributeValue(cu, DW_AT_abstract_origin, form_value)) {
      DWARFDIE die = form_value.Reference();
      if (die) {
```
Both of these go back to the DWARFDIE class which will cause all of the DIEs to be parsed in the DWARFUnit anyway. So the peek function will work, it will just do a lot more work than you expected. Note that if the `DW_AT_specification` or `DW_AT_abstract_origin` attributes can represent a DIE from another compile unit of they use the `DW_FORM_ref_addr`, which can point across compile units. So it will take some more work to make the `DWARFUnit::PeekDIEName` work and be as efficient as needed.

Reply:
I am not saying that this works needs to be done, just pointing out that this might not be as efficient as you want it to be and might cause all DIEs to be parsed.